File: kr_stock_app/lib/brokers/kiwoom.dart
// 키움증권 REST API (국내주식).
import 'package:http/http.dart' as http;

import '../core/market.dart';
import 'broker.dart';

const _nativeMinutes = [1, 3, 5, 10, 15, 30, 45, 60];

double _signed(Object? v, [double fallback = 0]) {
  final s = (v ?? '').toString().replaceAll(',', '').trim();
  return s.isEmpty ? fallback : (double.tryParse(s) ?? fallback);
}

class KiwoomBroker implements Broker {
  final String appKey, secretKey;
  @override
  final bool live;
  final http.Client client;
  final TokenCache tokens;
  final RateLimiter limiter;
  String? _token;

  KiwoomBroker(this.appKey, this.secretKey,
      {this.live = false, http.Client? client, KeyValueStore? store, Duration rateInterval = const Duration(milliseconds: 250)})
      : client = client ?? http.Client(),
        tokens = TokenCache(store ?? MemoryStore()),
        limiter = RateLimiter(rateInterval) {
    if (appKey.isEmpty || secretKey.isEmpty) throw BrokerException('키움 App Key / Secret Key 를 입력하세요');
  }

  @override
  String get name => 'kiwoom';
  String get baseUrl => live ? 'https://api.kiwoom.com' : 'https://mockapi.kiwoom.com';
  String get _tokenKey => 'kiwoom_${live}_${appKey.hashCode}';

  Future<String> _accessToken() async {
    if (_token != null) return _token!;
    final cached = await tokens.load(_tokenKey);
    if (cached != null) return _token = cached;
    final r = await sendJson(client, 'POST', Uri.parse('$baseUrl/oauth2/token'),
        headers: {'content-type': 'application/json;charset=UTF-8'},
        body: {'grant_type': 'client_credentials', 'appkey': appKey, 'secretkey': secretKey});
    final token = r.body['token'] as String?;
    if (token == null) throw BrokerException('키움 토큰 발급 실패: ${r.body['return_msg'] ?? r.body}');
    var ttl = 12 * 3600;
    final exp = '${r.body['expires_dt'] ?? ''}';
    if (exp.length == 14) {
      final t = parseYmdHms(exp.substring(0, 8), exp.substring(8));
      ttl = t.difference(nowKst()).inSeconds;
    }
    await tokens.save(_tokenKey, token, ttl);
    return _token = token;
  }

  Future<JsonResponse> _post(String path, String apiId, Map<String, dynamic> body,
      {String contYn = 'N', String nextKey = ''}) async {
    await limiter.wait();
    final r = await sendJson(client, 'POST', Uri.parse('$baseUrl$path'), headers: {
      'content-type': 'application/json;charset=UTF-8',
      'authorization': 'Bearer ${await _accessToken()}',
      'api-id': apiId,
      'cont-yn': contYn,
      'next-key': nextKey,
    }, body: body);
    if ('${r.body['return_code'] ?? 0}' != '0') {
      throw BrokerException('키움 $apiId 실패 [${r.body['return_code']}] ${r.body['return_msg'] ?? ''}');
    }
    return r;
  }

  @override
  Future<Quote> getQuote(String symbol) async {
    final info = await _post('/api/dostk/stkinfo', 'ka10001', {'stk_cd': symbol});
    final price = parseNum(info.body['cur_prc']);
    double? bid, ask;
    try {
      final b = await _post('/api/dostk/mrkcond', 'ka10004', {'stk_cd': symbol});
      ask = parseNum(b.body['sel_fpr_bid']);
      bid = parseNum(b.body['buy_fpr_bid']);
    } on BrokerException {
      // 호가 실패 시 스프레드 필터만 생략
    }
    return Quote(symbol, price, bid: bid == 0 ? null : bid, ask: ask == 0 ? null : ask);
  }

  @override
  Future<List<Candle>> getCandles(String symbol, String interval, int count) async {
    final minutes = minutesOf(interval);
    if (minutes == null) {
      return _chart(symbol, 'ka10081', {'base_dt': ymd(nowKst())}, 'stk_dt_pole_chart_qry', 'dt', count);
    }
    final native = _nativeMinutes.contains(minutes) ? minutes : 1;
    final c = await _chart(symbol, 'ka10080', {'tic_scope': '$native'}, 'stk_min_pole_chart_qry', 'cntr_tm',
        count * minutes ~/ native);
    final res = resampleMinutes(c, native == 1 ? minutes : 1);
    return res.length > count ? res.sublist(res.length - count) : res;
  }

  Future<List<Candle>> _chart(String symbol, String apiId, Map<String, String> extra, String listKey, String timeKey,
      int count) async {
    final m = <DateTime, Candle>{};
    var cont = 'N', key = '';
    for (var i = 0; i < 10; i++) {
      final r = await _post('/api/dostk/chart', apiId, {'stk_cd': symbol, 'upd_stkpc_tp': '1', ...extra},
          contYn: cont, nextKey: key);
      for (final row in ((r.body[listKey] as List?) ?? []).cast<Map>()) {
        final ts = '${row[timeKey] ?? ''}';
        if (ts.length < 8) continue;
        final t = ts.length >= 14 ? parseYmdHms(ts.substring(0, 8), ts.substring(8, 14)) : parseYmd(ts);
        m[t] = Candle(t, parseNum(row['open_pric']), parseNum(row['high_pric']), parseNum(row['low_pric']),
            parseNum(row['cur_prc']), parseNum(row['trde_qty']));
      }
      cont = r.headers['cont-yn'] ?? 'N';
      key = r.headers['next-key'] ?? '';
      if (m.length >= count || cont != 'Y' || key.isEmpty) break;
    }
    return sortedTail(m, count);
  }

  @override
  Future<Balance> getBalance() async {
    final dep = await _post('/api/dostk/acnt', 'kt00001', {'qry_tp': '3'});
    final cash = _signed(dep.body['ord_alow_amt'], _signed(dep.body['entr']));
    final positions = <Position>[];
    var total = 0.0;
    var cont = 'N', key = '';
    for (var i = 0; i < 10; i++) {
      final r = await _post('/api/dostk/acnt', 'kt00018', {'qry_tp': '1', 'dmst_stex_tp': 'KRX'}, contYn: cont, nextKey: key);
      total = _signed(r.body['tot_evlt_amt'], total);
      for (final row in ((r.body['acnt_evlt_remn_indv_tot'] as List?) ?? []).cast<Map>()) {
        final qty = parseNum(row['rmnd_qty']).toInt();
        if (qty > 0) {
          positions.add(Position('${row['stk_cd']}'.replaceFirst(RegExp('^A'), ''), qty, parseNum(row['pur_pric']),
              name: '${row['stk_nm'] ?? ''}'.trim(), currentPrice: parseNum(row['cur_prc'])));
        }
      }
      cont = r.headers['cont-yn'] ?? 'N';
      key = r.headers['next-key'] ?? '';
      if (cont != 'Y' || key.isEmpty) break;
    }
    return Balance(cash, total, positions);
  }

  @override
  Future<OrderResult> placeOrder(String symbol, Side side, int qty, OrderType type, double price) async {
    if (qty <= 0) return const OrderResult(false, message: '수량이 0입니다');
    final market = type == OrderType.market;
    try {
      final r = await _post('/api/dostk/ordr', side == Side.buy ? 'kt10000' : 'kt10001', {
        'dmst_stex_tp': 'KRX',
        'stk_cd': symbol,
        'ord_qty': '$qty',
        'ord_uv': market ? '' : '${price.round()}',
        'trde_tp': market ? '3' : '0',
        'cond_uv': '',
      });
      return OrderResult(true, orderId: '${r.body['ord_no'] ?? ''}', message: '${r.body['return_msg'] ?? ''}', price: price);
    } on BrokerException catch (e) {
      return OrderResult(false, message: e.message);
    }
  }
}

File: kr_stock_trader/brokers/kiwoom.py
"""키움증권 REST API 국내주식 클라이언트 (2025년 공개된 REST 방식, 영웅문 OCX 불필요).

- 실전: https://api.kiwoom.com
- 모의: https://mockapi.kiwoom.com
문서: https://openapi.kiwoom.com (API ID 기준: ka10001, ka10004, ka10080, ka10081, kt00001, kt00018, kt10000, kt10001)
"""

from datetime import datetime
from typing import List

import requests

from . import _token_cache
from .base import Broker, BrokerError, RateLimiter, parse_interval, resample_minutes
from ..market import KST, now_kst
from ..models import Balance, Candle, OrderResult, OrderType, Position, Quote, Side

BASE_URLS = {"real": "https://api.kiwoom.com", "demo": "https://mockapi.kiwoom.com"}
NATIVE_MINUTES = (1, 3, 5, 10, 15, 30, 45, 60)


def _num(v, default=0.0) -> float:
    """키움은 '+70000', '-70000', '000000070000' 같은 문자열로 준다. 가격 부호는 등락 표시라 절댓값."""
    if v in (None, ""):
        return default
    try:
        return abs(float(str(v).replace(",", "")))
    except ValueError:
        return default


def _signed(v, default=0.0) -> float:
    try:
        return float(str(v).replace(",", "")) if v not in (None, "") else default
    except ValueError:
        return default


class KiwoomBroker(Broker):
    name = "kiwoom"

    def __init__(self, app_key: str, secret_key: str, env: str = "demo",
                 session: requests.Session = None, rate_interval: float = 0.25):
        if env not in BASE_URLS:
            raise ValueError("env 는 demo 또는 real")
        if not app_key or not secret_key:
            raise BrokerError("KIWOOM_APP_KEY / KIWOOM_SECRET_KEY 가 필요합니다")
        self.app_key, self.secret_key, self.env = app_key, secret_key, env
        self.base_url = BASE_URLS[env]
        self.session = session or requests.Session()
        self.limiter = RateLimiter(rate_interval)
        self._token = None

    def _access_token(self) -> str:
        if self._token:
            return self._token
        ns = f"kiwoom_{self.env}"
        cached = _token_cache.load(ns, self.app_key)
        if cached:
            self._token = cached
            return cached
        r = self.session.post(f"{self.base_url}/oauth2/token", json={
            "grant_type": "client_credentials", "appkey": self.app_key, "secretkey": self.secret_key,
        }, headers={"content-type": "application/json;charset=UTF-8"}, timeout=10)
        data = r.json() if r.content else {}
        token = data.get("token")
        if not token:
            raise BrokerError(f"키움 토큰 발급 실패: {data.get('return_msg') or data or r.status_code}")
        try:
            expires = datetime.strptime(data["expires_dt"], "%Y%m%d%H%M%S").replace(tzinfo=KST).timestamp()
        except (KeyError, ValueError):
            expires = now_kst().timestamp() + 12 * 3600
        _token_cache.save(ns, self.app_key, token, expires)
        self._token = token
        return token

    def _post(self, path: str, api_id: str, body: dict, cont_yn: str = "N", next_key: str = ""):
        self.limiter.wait()
        headers = {
            "content-type": "application/json;charset=UTF-8",
            "authorization": f"Bearer {self._access_token()}",
            "api-id": api_id, "cont-yn": cont_yn, "next-key": next_key,
        }
        r = self.session.post(self.base_url + path, headers=headers, json=body, timeout=10)
        try:
            data = r.json()
        except ValueError:
            raise BrokerError(f"키움 응답 오류 HTTP {r.status_code}: {r.text[:200]}")
        if str(data.get("return_code", "0")) != "0":
            raise BrokerError(f"키움 {api_id} 실패 [{data.get('return_code')}] {data.get('return_msg', '')}")
        return data, r.headers.get("cont-yn", "N"), r.headers.get("next-key", "")

    # ---- 시세 -----------------------------------------------------------
    def get_quote(self, symbol: str) -> Quote:
        info, _, _ = self._post("/api/dostk/stkinfo", "ka10001", {"stk_cd": symbol})
        price = _num(info.get("cur_prc"))
        bid = ask = None
        try:
            book, _, _ = self._post("/api/dostk/mrkcond", "ka10004", {"stk_cd": symbol})
            ask, bid = _num(book.get("sel_fpr_bid")) or None, _num(book.get("buy_fpr_bid")) or None
        except BrokerError:
            pass
        return Quote(symbol, price, bid, ask)

    def get_candles(self, symbol: str, interval: str = "D", count: int = 100) -> List[Candle]:
        minutes = parse_interval(interval)
        if minutes is None:
            return self._chart(symbol, "ka10081", {"base_dt": now_kst().strftime("%Y%m%d")},
                               "stk_dt_pole_chart_qry", "dt", "%Y%m%d", count)
        native = minutes if minutes in NATIVE_MINUTES else 1
        candles = self._chart(symbol, "ka10080", {"tic_scope": str(native)},
                              "stk_min_pole_chart_qry", "cntr_tm", "%Y%m%d%H%M%S", count * minutes // native)
        return resample_minutes(candles, minutes if native == 1 else 1)[-count:]

    def _chart(self, symbol, api_id, extra, list_key, time_key, fmt, count):
        body = {"stk_cd": symbol, "upd_stkpc_tp": "1", **extra}
        candles, cont, key = {}, "N", ""
        for _ in range(10):
            data, cont, key = self._post("/api/dostk/chart", api_id, body, cont, key)
            for r in data.get(list_key) or []:
                if not r.get(time_key):
                    continue
                t = datetime.strptime(r[time_key], fmt).replace(tzinfo=KST)
                candles[t] = Candle(t, _num(r.get("open_pric")), _num(r.get("high_pric")),
                                    _num(r.get("low_pric")), _num(r.get("cur_prc")), _num(r.get("trde_qty")))
            if len(candles) >= count or cont != "Y" or not key:
                break
        return [candles[k] for k in sorted(candles)][-count:]

    # ---- 계좌 -----------------------------------------------------------
    def get_balance(self) -> Balance:
        dep, _, _ = self._post("/api/dostk/acnt", "kt00001", {"qry_tp": "3"})
        cash = _signed(dep.get("ord_alow_amt"), _signed(dep.get("entr")))
        positions, total, cont, key = [], 0.0, "N", ""
        for _ in range(10):
            data, cont, key = self._post("/api/dostk/acnt", "kt00018",
                                         {"qry_tp": "1", "dmst_stex_tp": "KRX"}, cont, key)
            total = _signed(data.get("tot_evlt_amt"), total)
            for r in data.get("acnt_evlt_remn_indv_tot") or []:
                qty = int(_num(r.get("rmnd_qty")))
                if qty > 0:
                    code = str(r.get("stk_cd", "")).lstrip("A")
                    positions.append(Position(code, qty, _num(r.get("pur_pric")),
                                              r.get("stk_nm", "").strip(), _num(r.get("cur_prc"))))
            if cont != "Y" or not key:
                break
        return Balance(cash=cash, total_eval=total, positions=positions)

    # ---- 주문 -----------------------------------------------------------
    def place_order(self, symbol, side, qty, order_type=OrderType.MARKET, price=0) -> OrderResult:
        if qty <= 0:
            return OrderResult(False, message="수량이 0입니다", symbol=symbol, side=side)
        is_market = order_type == OrderType.MARKET
        body = {
            "dmst_stex_tp": "KRX", "stk_cd": symbol, "ord_qty": str(int(qty)),
            "ord_uv": "" if is_market else str(int(price)),
            "trde_tp": "3" if is_market else "0", "cond_uv": "",
        }
        api_id = "kt10000" if side == Side.BUY else "kt10001"
        try:
            data, _, _ = self._post("/api/dostk/ordr", api_id, body)
        except BrokerError as e:
            return OrderResult(False, message=str(e), symbol=symbol, side=side, qty=qty, price=price)
        return OrderResult(True, str(data.get("ord_no", "")), data.get("return_msg", ""),
                           symbol, side, qty, price, raw=data)
